python/variable_init.cpp

```diff
@@ -0,0 +1,456 @@
+          return element_array<T>();
+        } else {
+          const auto [actual_unit, conversion_factor] =
+              common_unit<T>(value, unit);
+          unit = actual_unit;
+          if (conversion_factor != 1) {
+            // TODO Triggered once common_unit implements conversions.
+            std::terminate();
+          }
+          return element_array<T>(1, extract_scalar<T>(value, unit));
+        }
+      }();
+      auto var = [&]() {
+        if (variance.is_none()) {
+          return element_array<T>();
+        } else {
+          // No need to call common_unit as of now as it only matters for
+          // datetimes which cannot have variances.
+          return element_array<T>(1, extract_scalar<T>(variance, unit));
+        }
+      }();
+      if ((with_variance.has_value() && *with_variance) ||
+          (!with_variance.has_value() && var)) {
+        return makeVariable<T>(Values(std::move(val)),
+                               Variances(std::move(var)));
+      } else {
+        return makeVariable<T>(Values(std::move(val)));
+      }
+    }();
+
+    variable.setUnit(unit);
+    return variable;
+  }
+};
+
+Variable make_variable_scalar(const py::object &value,
+                              const py::object &variance,
+                              const units::Unit unit, DType dtype,
+                              const std::optional<bool> with_variance) {
+  dtype = common_dtype(value, variance, dtype, false);
+  return core::CallDType<
+      double, float, int64_t, int32_t, bool, scipp::core::time_point,
+      std::string, Variable, DataArray, Dataset,
+      python::PyObject>::apply<MakeVariableScalar>(dtype, value, variance, unit,
+                                                   with_variance);
+}
+
+template <class T> struct MakeVariableArray {
+  static Variable apply(const Dimensions &dims, const py::object &values,
+                        const py::object &variances, const units::Unit unit,
+                        const std::optional<bool> with_variance) {
+    auto variable =
+        ((with_variance.has_value() && *with_variance) ||
+         (!with_variance.has_value() && !variances.is_none()))
+            ? makeVariable<T>(
+                  dims, Values(dims.volume(), core::default_init_elements),
+                  Variances(dims.volume(), core::default_init_elements))
+            : makeVariable<T>(
+                  dims, Values(dims.volume(), core::default_init_elements));
+
+    const auto [actual_unit, conversion_factor] = common_unit<T>(values, unit);
+    variable.setUnit(actual_unit);
+    if (conversion_factor != 1) {
+      // TODO Triggered once common_unit implements conversions.
+      std::terminate();
+    }
+
+    if (!values.is_none()) {
+      copy_array_into_view(cast_to_array_like<T>(values, actual_unit),
+                           variable.template values<T>(), dims);
+    }
+    if (with_variance.value_or(true) && !variances.is_none()) {
```

Comment:
Does this mean we can provide `variances`, but ignore them using the `with_variance` arg?

Reply:
Yes

Reply:
Let's not do this? Just throw?